File: lang/cs/Org.Apache.REEF.DistributedR/Network/Serializer.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using Org.Apache.REEF.Utilities.Logging;
using Microsoft.Hadoop.Avro;

namespace Org.Apache.REEF.DistributedR.Network
{
    /// <summary>
    /// Serializes and desearializes distributed R messages with a header
    /// that identifies the message type. All messages types are automatically
    /// registered via type reflection.
    /// </summary>
    public static class Serializer
    {
        private static readonly Logger Logr = Logger.GetLogger(typeof(TaskRunner));

        private delegate void Serialize(MemoryStream stream, IMessage message);
        private delegate void Deserialize(MemoryStream stream, object observer);

        private static SortedDictionary<Guid, Serialize> serializeMap = new SortedDictionary<Guid, Serialize>();
        private static SortedDictionary<Guid, Deserialize> deserializeMap = new SortedDictionary<Guid, Deserialize>();

        /// <summary>
        /// Use reflection to find and register all messages derived from the client class.
        /// </summary>
        static Serializer()
        {
            Assembly assembly = typeof(Serializer).Assembly;
            Logr.Log(Level.Info, string.Format("Retrieving types for assembly: {0}", assembly.FullName));
            Type[] types = assembly.GetTypes();
            foreach (Type type in types)
            {
                if (type.FullName.ToLower().Contains("distributedr.network"))
                {
                    Type baseType = type.BaseType;
                    while (baseType != null)
                    {
                        if (baseType.Name.ToLower().Contains("serializerclient"))
                        {
                            Logr.Log(Level.Info, string.Format("Found type: {0}", type.FullName));
                            MethodInfo methodInfo = baseType.GetMethod(
                                "Initialize", BindingFlags.IgnoreCase | BindingFlags.Static | BindingFlags.NonPublic);
                            if (methodInfo != null)
                            {
                                methodInfo.Invoke(null, null);
                            }
                        }
                        baseType = baseType.BaseType;
                    }
                }
            }
        }

        /// <summary>
        /// Generates and stores the metadata necessary to serialze
        /// deserialize a specific message type.
        /// </summary>
        /// <typeparam name="TMessage"></typeparam>
        internal static void Register<TMessage>()
        {
            Type type = typeof(TMessage);
            Guid guid = GetGuid(type);

            Serialize serialize = (MemoryStream stream, IMessage message) =>
            {
                IAvroSerializer<TMessage> messageWriter = AvroSerializer.Create<TMessage>();
                messageWriter.Serialize(stream, (TMessage)message);
            };
            serializeMap.Add(guid, serialize);

            Deserialize deserialize = (MemoryStream stream, object observer) =>
            {
                IAvroSerializer<TMessage> messageReader = AvroSerializer.Create<TMessage>();
                TMessage message = messageReader.Deserialize(stream);
                IObserver<TMessage> msgObserver = observer as IObserver<TMessage>;
                if (msgObserver != null)
                {
                    msgObserver.OnNext(message);
                }
                else
                {
                    Logr.Log(Level.Info, string.Format("Unhandled message received: " + message.ToString()));
                }
            };
            deserializeMap.Add(guid, deserialize);
        }

        /// <summary>
        /// Retrieves the GUID attribute for the specified type.
        /// </summary>
        /// <param name="type">An instance of System.Type.</param>
        /// <returns>Guid</returns>
        private static Guid GetGuid(Type type)
        {
            GuidAttribute attribute = (GuidAttribute)Attribute.GetCustomAttribute(type, typeof(GuidAttribute));
            if (attribute != null)
            {
                return new Guid(attribute.Value);
            }
            else
            {
                throw new Exception("Distributed R messages must have a Guid attribute");
            }
        }

        /// <summary>
        /// Serialize the input message and return a byte array.
        /// </summary>
        /// <param name="message">A messeage to be serialized which derives from Client<TMessage></TMessage></param>
        /// <returns>A byte array containing the serialized message and header.</returns>
        public static byte[] Write(IMessage message) 
        {
            try
            { 
                IAvroSerializer<Header> headWriter = AvroSerializer.Create<Header>();
                using (MemoryStream stream = new MemoryStream())
                {
                    Guid id = GetGuid(message.GetType());
                    Header header;
                    header.Id = id;
                    headWriter.Serialize(stream, header);

                    Serialize serialize;
                    if (serializeMap.TryGetValue(id, out serialize))
                    {
                        serialize(stream, message);
                    }
                    else
                    {
                        throw new Exception("Request to serialize unknown message type.");
                    }
                    return stream.GetBuffer();
                }
            }
            catch (Exception e)
            {
                Logr.Log(Level.Error, "Failure writing message: " + e.Message);
                throw e;
            }
        }

        /// <summary>
        /// Read a message from the input byte array.
        /// </summary>
        /// <typeparam name="TReceiver">The concrete type of the object which implements the IReceiver interface.</typeparam>
        /// <param name="data">The byte array containing the message to be deserialized.</param>
        /// <param name="observer">An instance of the TReceiver object whose methods will be invoked base on the message type.</param>
        public static void Read(byte[] data, object observer)
        {
            try
            {
                IAvroSerializer<Header> headReader = AvroSerializer.Create<Header>();
                using (MemoryStream stream = new MemoryStream(data))
                {
                    Header head = headReader.Deserialize(stream);
                    Deserialize deserialize;
                    if (deserializeMap.TryGetValue(head.Id, out deserialize))
                    {
                        deserialize(stream, observer);
                    }
                    else
                    {
                        throw new Exception("Request to deserialize unknown message type.");
                    }
                }
            }
            catch (Exception e)
            {
                Logr.Log(Level.Error, "Failure reading message: " + e.Message);
                throw e;
            }
        }
    }
}

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/bridge/Serializer.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.reef.bridge;

import org.apache.avro.io.*;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.avro.specific.SpecificDatumWriter;
import org.apache.avro.specific.SpecificRecord;
import org.apache.reef.bridge.message.Header;
import org.apache.avro.specific.SpecificRecordBase;
import org.apache.reef.wake.rx.Observer;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.logging.Level;
import java.util.logging.Logger;
import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
import io.github.lukehutch.fastclasspathscanner.scanner.ScanResult;

/**
 * Base interface for anonymous message serializer objects.
 */
interface MessageSerializer {
  void serialize(ByteArrayOutputStream outputStream, SpecificRecord message) throws IOException;
}

/**
 * Abstract implementation of message serializer that carries the message type.
 * @param <TMessage> The type of message the instantiation can serialize.
 */
abstract class GenericMessageSerializer<TMessage> implements MessageSerializer {
  Class<TMessage> msgMetaClass;
  public GenericMessageSerializer(Class<TMessage> msgMetaClass) {
    this.msgMetaClass = msgMetaClass;
  }
  abstract public void serialize(ByteArrayOutputStream outputStream, SpecificRecord message) throws IOException;
}

/**
 * Base interface for ananymous message deserializer objects.
 */
interface MessageDeserializer {
  void deserialize(BinaryDecoder decoder, Object observer) throws IOException;
}

/**
 * Abstract implementation of message deserializer that carries the message type.
 * @param <TMessage> The type of message the instantiation can deserialize.
 */
abstract class GenericMessageDeserializer<TMessage> implements  MessageDeserializer {
  Class<TMessage> msgMetaClass;
  public GenericMessageDeserializer(Class<TMessage> msgMetaClass) { this.msgMetaClass = msgMetaClass; }
  abstract public void deserialize(BinaryDecoder decoder, Object observer) throws IOException;
}

/**
 *
 */
final public class Serializer {
  private static final Logger LOG = Logger.getLogger(Serializer.class.getName());
  // Maps for mapping message class names to serializer and deserializer classes.
  private static Map<String, MessageSerializer> nameToSerializerMap = new HashMap<>();
  private static Map<String, MessageDeserializer> nameToDeserializerMap = new HashMap<>();

  /**
   *  Class utility with all static methods.
   */
  private Serializer() { }

  /**
   * Finds all of the messages in the org.apache.reef.bridge.message package and calls register.
   */
  public static void Initialize() {
    LOG.log(Level.INFO, "Start: Serializer.Initialize");

    // Build a list of the message reflection classes.
    final ScanResult scanResult = new FastClasspathScanner("org.apache.reef.bridge.message").scan();
    final List<String> scanNames = scanResult.getNamesOfSubclassesOf(SpecificRecordBase.class);
    final List<Class<?>> messageClasses = scanResult.classNamesToClassRefs(scanNames);
    LOG.log(Level.INFO, "!!!NUMBER OF MESSAGES = " + Integer.toString(messageClasses.size()));

    try {
      // Register all of the messages.
      for (Class<?> cls : messageClasses) {
        LOG.log(Level.INFO,"Found message class: " + cls.getName() + " " + cls.getSimpleName());
        Method register = Serializer.class.getMethod("Register", cls.getClass());
        LOG.log(Level.INFO,"Obtained the method class instance");
        register.invoke(null, cls);
      }
    } catch (Exception e) {
      LOG.log(Level.SEVERE,"Failed to register message class " + e.getMessage() );
    }

    LOG.log(Level.INFO,"End: Serializer.Initialize");
  }

  /**
   * Instantiates and adds a message serializer/deserializer for the message.
   * @param msgMetaClass The reflection class for the message.
   * @param <TMessage> The message Java type.
   */
  public static <TMessage> void Register(Class<TMessage> msgMetaClass)
  {
    LOG.log(Level.INFO,"Registering [" + msgMetaClass.getSimpleName() + "]");

    // Instantiate an anonymous instance of the message serializer for this message type.
    final MessageSerializer messageSerializer = new GenericMessageSerializer<TMessage>(msgMetaClass) {

      public void serialize(ByteArrayOutputStream outputStream, SpecificRecord message) throws IOException {
        // Binary encoder for both the header and message.
        final BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(outputStream, null);

        // Writers for header and message.
        final DatumWriter<Header> headerWriter = new SpecificDatumWriter<>(Header.class);
        final DatumWriter<TMessage> messageWriter = new SpecificDatumWriter<>(msgMetaClass);

        // Write the header and the message.
        headerWriter.write(new Header(0, msgMetaClass.getSimpleName()), encoder);
        messageWriter.write((TMessage)message, encoder);
        encoder.flush();
      }

    };
    nameToSerializerMap.put(msgMetaClass.getSimpleName(), messageSerializer);

    // Instantiate an anonymous instance of the message deserializer for this message type.
    final MessageDeserializer messageDeserializer = new GenericMessageDeserializer<TMessage>(msgMetaClass) {

      public void deserialize(BinaryDecoder decoder, Object observer) throws IOException {
        final SpecificDatumReader<TMessage> messageReader = new SpecificDatumReader<>(msgMetaClass);
        final TMessage message = messageReader.read(null, decoder);
        if (observer instanceof Observer) {
          ((Observer<TMessage>)observer).onNext(message);
        }
      }

    };
    nameToDeserializerMap.put(msgMetaClass.getSimpleName(), messageDeserializer);
  }

  /**
   *
   * @param message
   */
  public static byte[] write(SpecificRecord message) {
    try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
      final String name = message.getClass().getSimpleName();
      LOG.log(Level.INFO, "Serialing message: [" + name + "]");

      final MessageSerializer serializer = nameToSerializerMap.get(name);
      serializer.serialize(outputStream, message);

      return(outputStream.toByteArray());
    } catch (Exception e) {
      LOG.log(Level.SEVERE, "Failed to serialize avro message: " + e.getMessage());
    }
  }

  /**
   *
   * @param messageBytes
   * @param observer
   */
  public static void read(byte[] messageBytes, Object observer) {
    try (InputStream inputStream = new ByteArrayInputStream(messageBytes)) {
      // Binary decoder for both the header and the message.
      final BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(inputStream, null);

      // Read the header message.
      final SpecificDatumReader<Header> headerReader = new SpecificDatumReader<>(Header.class);
      final Header header = headerReader.read(null, decoder);
      LOG.log(Level.INFO, "Deserializing message: [" + header.getClassName() + "]");

      // Get the appropriate deserializer and deserialize the message.
      final MessageDeserializer deserializer = nameToDeserializerMap.get(header.getClassName());
      deserializer.deserialize(decoder, observer);

    } catch (Exception e) {
      LOG.log(Level.SEVERE, "Failed to deserialize avro message: " + e.getMessage());
    }
  }
}
